ruby_on_rails/carrierwave.md

```diff
@@ -0,0 +1,216 @@
+Carrierwave is used to upload files to S3.
+
+1. Add `gem 'carrierwave'` and `gem 'fog-aws'`
+1. In `environment.rb` add `require 'carrierwave/orm/activerecord'`
+1. Add to your `application.example.yml`:
+
+```
+#CARRIERWAVE_SALT: 'rake secret'
+#S3_ACCESS_KEY_ID: ''
+#S3_ASSET_HOST: ''
+#S3_BUCKET_NAME: ''
+#S3_SECRET_ACCESS_KEY: ''
+```
+
+4. Add `initializers/carrierwave.rb`
+
+```rb
+CarrierWave.configure do |config|
+  if Rails.env.test?
+    config.storage = :file
+    config.enable_processing = false
+  elsif ENV['S3_BUCKET_NAME'].present?
+    config.fog_provider = 'fog/aws'
+    config.fog_credentials = {
+      provider: 'AWS',
+      aws_access_key_id: ENV['S3_ACCESS_KEY_ID'],
+      aws_secret_access_key: ENV['S3_SECRET_ACCESS_KEY'],
+      region: 'eu-central-1'
+    }
+    config.fog_directory = ENV['S3_BUCKET_NAME']
+    config.fog_public = true
+    config.fog_attributes = { cache_control: "public, max-age=#{365.days.to_i}" }
+    config.storage = :fog
+    config.asset_host = ENV['S3_ASSET_HOST']
+  else
+    config.storage = :file
+  end
+end
+```
+
+5. The `UploaderBasepath` is used, so we don't polute the `public/uploader` folder in tests. Therefore you have to add this:
+
+```rb
+module UploaderBasepath
+  extend ActiveSupport::Concern
+
+  included do
+    private
+
+    def base_path_helper
+      return "tmp/uploads" if Rails.env.test?
+      'uploads'
+    end
+  end
+end
+```
+
+And also we want to clear the folder after the tests run in a `spec/support/carrierwave.rb`
+
+```rb
+# frozen_string_literal: true
+
+uploads_test_path = Rails.root.join('public', 'tmp')
+
+RSpec.configure do |config|
+  config.after(:suite) do
+    FileUtils.rm_rf(Dir[uploads_test_path])
+  end
+end
+```
+
+6. The `SecurelyUploadable` makes sure, that your path is not guessable:
+
+```rb
+module SecurelyUploadable
+  extend ActiveSupport::Concern
+  include UploaderBasepath
+
+  included do
+    def store_dir
+      "#{base_path_helper}/#{model.class.to_s.underscore}/#{mounted_as}/#{unguessable_id}"
+    end
+
+    private
+
+    def unguessable_id
+      secret = [ENV['CARRIERWAVE_SALT'], model.id].join('/')
+      Digest::SHA256.hexdigest(secret)
+    end
+  end
+end
+```
+
+7. Implement an uploader e.g.
+
+```rb
+class PictureUploader < CarrierWave::Uploader::Base
+  include SecurelyUploadable
+
+  # add different sizes like that:
+  # version :preview do
+  #   process resize_to_fill: [100, 100]
+  # end
+end
+```
+
+8. We recommend to not run the image processing for all kind of specs, since it's slow. Therefore the initializer sets it to `false`. But for system specs we want to enable it.
+
+* seeds_spec.rb: Not really relevant, but also often a system spec
+
+```rb
+  around do |example|
+    described_class.enable_processing = false
+    example.run
+    described_class.enable_processing = true
+  end
+```
+
+* all specs except `system` specs:
+
+```rb
+config.before(:all, type: :system) do
+  PictureUploader.enable_processing = true
```

Comment:
@coorasse this is the plan:

1. disable it in general
2. enable it for `system`-specs
3. you might disable it for specific system specs e.g. `seeds_spec.rb`

So i'd need to do something like that, right:

```rb
config.before(:each, type: :system) do
  around do |example|
    described_class.enable_processing = true
    example.run
    described_class.enable_processing = false
  end
end

Reply:
I agree with the plan, but here you are enabling only one uploader. You probably want to do something like:
`CarrierWave.config.enable_processing = true` and not enabling a specific uploader. if you have more uploaders, you have to remember to enable or disable all of them

Reply:
I guess it would anyways be better to only allow opt-in into processing. Makes faster tests. You normally only need processing on if you are testing the processing itself. I'd disable it per default.